### PR TITLE
Migrated to OpenZepplin 1.9.0

### PR DESCRIPTION
**Pull request by vsavovlime**

### CODE REVIEW

File: contracts/Crowdsale/BasicCrowdsale.sol [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    constructor(uint256 _rate, address _wallet, address _token, uint256 _openingTime, uint256 _closingTime, uint256 _cap)
+    Crowdsale(_rate, _wallet, ERC20(_token))
+    TimedCrowdsale(_openingTime, _closingTime)
+    CappedCrowdsale(_cap) public {
+        require((_closingTime-_openingTime) > (15 * 1 days));
+    }
+
+    function buyTokens(address beneficiary) public payable {
+        require(msg.value >= MIN_CONTRIBUTION_AMOUNT);
```
Comment (Perseverance) [inline contracts/Crowdsale/BasicCrowdsale.sol]: Let's move this require to _preValidatePurchase

File: contracts/Crowdsale/BasicCrowdsale.sol [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    function _getTokenAmount(uint256 _weiAmount) internal view returns (uint256) {
+        uint256 _rate = getRate();
+        return _weiAmount.mul(_rate);
+    }
+
+    function createBountyToken(address beneficiary, uint256 amount) public onlyOwner returns (bool) {
+        require(!hasClosed());
+        MintableToken(token).mint(beneficiary, amount);
+        LogBountyTokenMinted(msg.sender, beneficiary, amount);
```
Comment (Perseverance) [inline contracts/Crowdsale/BasicCrowdsale.sol]: I think we should have `emit` before the events

File: contracts/Crowdsale/Vesting.sol [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    function claim() public payable onlyOwner {
+        ERC20 token = ERC20(tokenAddress);
+        require(token.balanceOf(this) > 0);
+        require(now > startDate);
+
+        if (tokenBalance == 0) {
+            tokenBalance = token.balanceOf(this);
+            uint256 owedFirstPeriodTokens = (tokenBalance * 20) / 100;
```
Comment (Perseverance) [inline contracts/Crowdsale/Vesting.sol]: there should not be uint256 here as you will be overriding the variables

File: contracts/Crowdsale/BasicCrowdsale.sol [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+contract BasicCrowdsale is MintedCrowdsale, FinalizableCrowdsale, CappedCrowdsale {
+
+    uint256 constant MIN_CONTRIBUTION_AMOUNT = 100 finney;
+    
+    uint256 constant PRESALE_CAP = 850 ether;
+    uint256 constant PRESALE_RATE = 6750;
+    uint256 constant PRESALE_DURATION = 14 days;
+
+    uint256 constant BONUS_1_CAP = PRESALE_CAP + 1700 ether;
```
Comment (Perseverance) [inline contracts/Crowdsale/BasicCrowdsale.sol]: Let's move the 1700 ether to constant variable

File: backend-scripts/crowdsale-services.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+class CrowdsaleService {
+	static async getOwner() {
+		return await WhitelistedRefundableCrowdsaleContract.owner();
+	}
+
+	static async getCurrentTokensForEther() {
+		const etherInWei = 1000000000000000000; // 1 ETH
+		let rate = await WhitelistedRefundableCrowdsaleContract.getRate();
+		const tokens = rate * etherInWei;
```
Comment (Perseverance) [inline backend-scripts/crowdsale-services.js]: If the rate becomes too big the JS will have a precision error. please use `ethers.utils.bignumberify`. 

https://docs.ethers.io/ethers.js/html/api-utils.html#creating-instances

File: backend-scripts/crowdsale-services.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		return isWhitelisted;
+	}
+
+	static async whitelistAddress(privateKey, address) {
+		let wallet = await WalletService.initWalletFromPrivateKey(privateKey);
+
+		let WhitelistedRefundableCrowdsaleContractWallet = WhitelistedRefundableCrowdsaleContractWithWallet(wallet);
+		const options = { gasPrice: constants.gasPrice };
+
+		return await WhitelistedRefundableCrowdsaleContractWallet.addToWhitelist(address, options);
```
Comment (Perseverance) [inline backend-scripts/crowdsale-services.js]: We should add waitForTransaction here and in the multiple whitelist:
https://docs.ethers.io/ethers.js/html/api-providers.html#waiting-for-transactions

Example:
```
const result = await registryContract.acceptNextJob();
await provider.waitForTransaction(result.hash);
```

File: contracts/Crowdsale/BasicCrowdsale.sol (hunk 1 above)
Comment (vsavovlime) [reply contracts/Crowdsale/BasicCrowdsale.sol]: Done

File: contracts/Crowdsale/BasicCrowdsale.sol (hunk 2 above)
Comment (vsavovlime) [reply contracts/Crowdsale/BasicCrowdsale.sol]: Done

File: contracts/Crowdsale/Vesting.sol (hunk 3 above)
Comment (vsavovlime) [reply contracts/Crowdsale/Vesting.sol]: Done

File: contracts/Crowdsale/BasicCrowdsale.sol (hunk 4 above)
Comment (vsavovlime) [reply contracts/Crowdsale/BasicCrowdsale.sol]: Done

File: backend-scripts/crowdsale-services.js (hunk 5 above)
Comment (vsavovlime) [reply backend-scripts/crowdsale-services.js]: Done

File: backend-scripts/crowdsale-services.js (hunk 6 above)
Comment (vsavovlime) [reply backend-scripts/crowdsale-services.js]: Done

**Pull request merged**